api/api_public_blockchain.go

```diff
@@ -380,28 +381,53 @@ func (s *PublicBlockChainAPI) DoEstimateGas(ctx context.Context, b Backend, args
 	cap = hi
 
 	// Create a helper to check if a gas allowance results in an executable transaction
-	executable := func(gas uint64) bool {
+	executable := func(gas uint64) (bool, []byte, error, error) {
 		args.Gas = hexutil.Uint64(gas)
-
-		_, _, _, failed, err := DoCall(ctx, b, args, rpc.NewBlockNumberOrHashWithNumber(rpc.LatestBlockNumber), vm.Config{UseOpcodeComputationCost: true}, localTxExecutionTime, gasCap)
-		if err != nil || failed {
-			return false
+		ret, _, _, status, err := DoCall(ctx, b, args, rpc.NewBlockNumberOrHashWithNumber(rpc.LatestBlockNumber), vm.Config{}, 0, gasCap)
+		if err != nil {
+			if errors.Is(err, blockchain.ErrIntrinsicGas) {
+				// Special case, raise gas limit
+				return false, ret, nil, nil
+			}
+			// Returns error when it is not VM error (less balance or wrong nonce, etc...).
+			return false, nil, nil, err
 		}
-		return true
+		// If err is vmError, return vmError with returned data
+		vmErr := blockchain.GetVMerrFromReceiptStatus(status)
+		if vmErr != nil {
+			return false, ret, vmErr, nil
+		}
+		return true, ret, vmErr, nil
 	}
 	// Execute the binary search and hone in on an executable gas limit
 	for lo+1 < hi {
 		mid := (hi + lo) / 2
-		if !executable(mid) {
+		isExecutable, _, _, err := executable(mid)
+		if err != nil {
+			return 0, err
+		}
+		if !isExecutable {
 			lo = mid
 		} else {
 			hi = mid
 		}
 	}
 	// Reject the transaction as invalid if it still fails at the highest allowance
 	if hi == cap {
-		if !executable(hi) {
-			return 0, fmt.Errorf("gas required exceeds allowance or always failing transaction")
+		isExecutable, ret, vmErr, err := executable(hi)
+		if err != nil {
+			return 0, err
+		}
+		if !isExecutable {
+			if vmErr != nil {
+				// Treat vmErr as RevertError only when there was returned data from call.
+				if isReverted(vmErr) && len(ret) > 0 {
+					return 0, newRevertError(ret)
+				}
+				return 0, vmErr
+			}
+			// Otherwise, the specified gas cap is too low
+			return 0, fmt.Errorf("gas required exceeds allowance (%d)", cap)
```

Comment:
As the comment above (misconsidered runtime gas usage (`constantGas` + `dynamicGas`), this routine seems to be reconsidered. If you filter the class of the gas error, you should correctly handle all types of gas error, not only the intrinsicGas. Furthermore, I recommend not making a big change to the existing function signature if the goal (awareness of the gas type error) is preserved. For example, `executable` function may have an additional return type of error to see that error type and its small change seems enough.